### PR TITLE
8325144: C1: Optimize CriticalEdgeFinder

### PR DESCRIPTION
**Pull request by D-D-H**
Hi,

Please help review this change that removed the sorting process in split_edges by checking if the 'to' block has been substituted.

Thanks

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8325144](https://bugs.openjdk.org/browse/JDK-8325144): C1: Optimize CriticalEdgeFinder (**Enhancement** - P4)

### Reviewers
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**) ⚠️ Review applies to [8491470d](https://git.openjdk.org/jdk/pull/17674/files/8491470d0e80bf85b1a8d659078d6946becf7d9c)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to [7d15411b](https://git.openjdk.org/jdk/pull/17674/files/7d15411bfc066d31e5924777401d4bae08aadfcf)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17674/head:pull/17674` \
`$ git checkout pull/17674`

Update a local copy of the PR: \
`$ git checkout pull/17674` \
`$ git pull https://git.openjdk.org/jdk.git pull/17674/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17674`

View PR using the GUI difftool: \
`$ git pr show -t 17674`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17674.diff">https://git.openjdk.org/jdk/pull/17674.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17674#issuecomment-1924930118)

### CODE REVIEW

**Comment by dean-long** [comment]
Sorry, could you explain why the new code is functionally equivalent to the old code?  It's not obvious to me.

**Comment by D-D-H** [comment]
The functionality of `CriticalEdgeFinder` is to insert a new block between the two blocks which satisfy the following conditions:

- the first block has at least 2 successors (the `from` block)
- the second block has at least 2 predecessors (the `to` block)
- the first is a predecessor of the second block

The flow of CriticalEdgeFinder:
1. collects all the block pairs whose `from` and `to` blocks satisfy the conditions above (`CriticalEdgeFinder::block_do`)
2. traverse the pairs to insert new blocks (`CriticalEdgeFinder::split_edges`)

In some cases (e.g. `LookupSwitch`), Several successors of a block can be the same. Hence, in `CriticalEdgeFinder::split_edges`, the pairs are sorted before handling them. And `last_pair` is used to check if a pair is handled already.

Since `BlockBegin::insert_block_between` replaces all matched successors with the new block and doesn't change the ordering or number of successors of the `from` block, we can check if the pair is handled by checking if the corresponding successor of the `from` block is changed. So we can remove the sorting process and simplify the code.

The only difference compared to the original result is that `_block_id` of new blocks changes. If I understand correctly, it doesn't influence subsequent compilation processes.

File: src/hotspot/share/c1/c1_IR.cpp
Comment (dean-long) [inline src/hotspot/share/c1/c1_IR.cpp:324]: Yes, we can remove the ctor, but wouldn't it be better to keep the ctor to initialize `blocks`?

File: src/hotspot/share/c1/c1_Instruction.hpp
Comment (dean-long) [inline src/hotspot/share/c1/c1_Instruction.hpp:2427]: We don't really need `_to` anymore, because sux_at(_index) == _to if `insert_block_between` hasn't already processed this edge.  We can detect if sux_at(_index)  has been replaced in a couple different ways:
1. set a block flag on the newly inserted block
2. check the block_id on the newly inserted block, assume block ids increase monotonically.
But maybe the added complexity isn't worth it?  What's the performance improvement with this change vs the footprint increase for this new field?

File: src/hotspot/share/c1/c1_IR.cpp
Comment (D-D-H) [reply src/hotspot/share/c1/c1_IR.cpp:324]: Agree. Updated

File: src/hotspot/share/c1/c1_Instruction.hpp
Comment (D-D-H) [reply src/hotspot/share/c1/c1_Instruction.hpp:2426]: Inspired by your comment.
Only the method `BlockBegin::insert_block_between` sets `critical_edge_split_flag`, so we can check if that flag is set or not.
Then there is no footprint problem.

**Review by dean-long — APPROVED** [review]
I like it!

**Comment by D-D-H** [comment]
Could I get a second review?

**Review by chhagedorn — APPROVED** [review]
Looks good!

> The functionality of CriticalEdgeFinder is to insert a new block between the two blocks which satisfy the following conditions:

While at it, maybe you also want to add a concise summary as a comment for the `CriticalEdgeFinder` class.

File: src/hotspot/share/c1/c1_IR.cpp
Comment (chhagedorn) [inline src/hotspot/share/c1/c1_IR.cpp:315]:
```suggestion
// (all new blocks set this flag) we can avoid repeated processing.
```

**Review by chhagedorn — APPROVED** [review]
Thanks for adding the comment!

**Comment by D-D-H** [comment]
@dean-long @chhagedorn Thanks for the review.

/integrate

**Pull request closed without merging**
